### DISCUSSION TITLE
Will be nice to have an option to save filter selection in session

### DISCUSSION THREAD

**Discussion by dgillier** [rappasoft/laravel-livewire-tables#1251] [Feature Requests]
I’m having multiple filters on multiple tables in multiple pages, and in most case I’m loosing my filter selection passed by url after some navigations.

So I’m wondering if having a table option to save filters selection in session will be possible ?
Thanks,
Denis

**Comment by lrljoe** [comment]
Can you give me a tighter example.

Are you using the same filter name across multiple tables?

Are you using the same tableName attribute on each component?

Once I can find the problem, more than happy to fix it.

I was considering (optional) browser side filter caching for v3, so you come back and it's exactly how you left it.

In the interim, the filter statuses etc are all stored in the attribute for the table name, so you can relatively easily push/pull from that on mount.

**Reply by dgillier** [reply]
Humm...
Here is my investigation:

I'm using a back button on some pages initially using the Laravel -> previous() function, but Livewire is not saving the url...
So  I tried to use a middleware solution (https://github.com/ralphjsmit/livewire-urls) pushing urls in a session queue, but for some unknown reason neither fullUrl() or fullUrlWithQuery() are saving the query constructed by Livewire...

So it's not a problem specific to this package but to Livewire (or my understanding of it).

It's why I'm thinking about storing filter selection in sessions, but maybe I'm missing something...

Tks,
Denis

**Reply by lrljoe** [reply]
LiveWire3 does make some tweaks by the looks of things on how this behaves.

I'm still going to take a look and try to replicate this though so that I can understand the issue, as it won't be an overnight transition to LiveWire3 when it comes out later this month.

Are you using any setFilter() in your mount or configure() methods?
Or setting defaults on the filter themselves?

A session approach would almost certainly work, but there'll be a cleaner way

**Reply by lrljoe** [reply]
@dgillier when you tried [ralphjsmit/livewire-urls](https://github.com/ralphjsmit/livewire-urls) - where did you add the middleware?  May be worth trying it in the Livewire middleware slot for persistence.

Worth experimenting with setting this to "true", in your config/livewire.php:
```
    'back_button_cache' => true,
```
You can of course set it on-the-fly for your components.  Caution that setting this to true globally can cause some issues with stale data being displayed.

You could of course also just update the session value yourself using a custom middleware, setting the back button to the referrer/HTTP_REFERER value.

I'm experimenting with some elements for you, but I'm trying to avoid overloading the session.  The Column Selection is stored, however, doing the same with filters, where you for example have a "default value" set, or are setting a default via configure() or similar, all makes life a little more complicated!

I'd assume that stored filter values would obviously over-ride default settings, so there's a fair few places that will need adjusting to cater for it.

I think safely assume it won't be this month, but I don't see why it can't be a future optional feature.

**Reply by lrljoe** [reply]
So I had a play, and have a working example, however there's a shed load of testing that I'll need to do before its even remotely ready, so I'm not even thinking about posting the code on here, and it's definitely not ready for widespread consumption

BUT if you jump on the Discord then I'll PM you an example, and can see what you think, as you're always so helpful in raising ideas/issues :)

**Reply by dgillier** [reply]
Thanks :)
I tried to connect to discord with the link in the discussion page, but got this:

![Screenshot 2023-07-12 at 13 07 38](https://github.com/rappasoft/laravel-livewire-tables/assets/37960204/7a35de4b-9fc9-4c79-99f1-ecbf67f87f8e)

**Reply by lrljoe** [reply]
[https://discord.gg/9dR8R3DwB](https://discord.gg/9dR8R3DwB)

Think I need to update the link!

**Reply by lrljoe** [reply]
So I've got a working prototype for persisting the filters, just need to look at how long it persists for, as the table fingerprint does change sporadically.

It's now configurable in my test environment (I.e. enable or disable the feature), and it seems to persist for a time, including hammering the back button and then browsing manually back to the correct page with the table.

I do need to test across all of the types of filters, but I don't see why it'd suddenly not work.

It'll probably be a month or so before it's in a release due to the tests and docs, but feel free to DM me on discord and I'll point you at which fork and branch it's currently in.

**Reply by lrljoe** [reply]
You'll be pleased to know that Tables v3 uses a slightly different approach (using Livewire 3), and the filters do actually track correctly with the back button.

I am still going to add the option to persist the filter values in the session, as it may be useful to someone!

**Comment by lrljoe** [comment]
It crossed my mind that you could actually do this within your table component if you have a pressing need by reading/writing from the variable that the filter values are stored in:
```
$this->{$this->tableName}['filters']
```
Not ideal, but will probably do for the interim while we finish off v3.

**Reply by dgillier** [reply]
Oh great :)
Thanks a lot, sorry, I was not very active this last month!

**Reply by lrljoe** [reply]
No problem, your absolute best bet would be to run it at boot, ideally in a trait to allow re-use.

There is actually code in the package for persisting it to a database, as when you use the reorder function, that is how it restores where you were.

Using that as a jumping off point, something like this would allow you to save the exact state of the table, which page, column select, the works:

```
        session([' **ADD A UNIQUE KEY HERE** ' => [
            $this->getTableName() => $this->{$this->getTableName()},
            'sorts' => $this->sorts,
            'search' => $this->search,
            'selectedColumns' => $this->selectedColumns,
            'sortingPillsStatus' => $this->getSortingPillsStatus(),
            'sortingStatus' => $this->getSortingStatus(),
            'paginationStatus' => $this->getPaginationStatus(),
            'perPageVisibilityStatus' => $this->getPerPageVisibilityStatus(),
            'perPageAccepted' => $this->getPerPageAccepted(),
            'perPage' => $this->getPerPage(),
            'page' => $this->paginators[$this->getComputedPageName()] ?? 1,
            'searchStatus' => $this->getSearchStatus(),
            'bulkActionsStatus' => $this->getBulkActionsStatus(),
            'selected' => $this->getSelected(),
            'selectAllStatus' => $this->getSelectAllStatus(),
            'filtersStatus' => $this->getFiltersStatus(),
            'secondaryHeaderStatus' => $this->getSecondaryHeaderStatus(),
            'footerStatus' => $this->getFooterStatus(),
            'collapsingColumnsStatus' => $this->hasCollapsingColumns(),
        ]]);
```
and then to restore:
```
        if (session()->has(' THE UNIQUE KEY YOU SET BEFORE ')) {
            $save = session()->get(' THE UNIQUE KEY YOU SET BEFORE ');
            $this->{$this->getTableName()} = $save[$this->getTableName()];
            $this->sorts = $save['sorts'];
            $this->search = $save['search'];
            $this->selectedColumns = $save['selectedColumns'];
            $this->setSortingPillsStatus($save['sortingPillsStatus']);
            $this->setSortingStatus($save['sortingStatus']);
            $this->setPaginationStatus($save['paginationStatus']);
            $this->setPerPageVisibilityStatus($save['perPageVisibilityStatus']);
            $this->setPerPageAccepted($save['perPageAccepted']);
            $this->setPerPage($save['perPage']);
            $this->setPage($save['page'], $this->getComputedPageName());
            $this->setSearchStatus($save['searchStatus']);
            $this->setBulkActionsStatus($save['bulkActionsStatus']);
            $this->setSelected($save['selected']);
            $this->setSelectAllStatus($save['selectAllStatus']);
            $this->setFiltersStatus($save['filtersStatus']);
            $this->setSecondaryHeaderStatus($save['secondaryHeaderStatus']);
            $this->setFooterStatus($save['footerStatus']);
            $this->setCollapsingColumnsStatus($save['collapsingColumnsStatus']);
            session()->forget(' THE UNIQUE KEY YOU SET BEFORE ');
   }
```

**Reply by dnielandt** [reply]
Seeing this peace of code gives me a great starting point for a simular problem. 
Just not sure wich function tracks the changes in filters, selectedColumns, ...

**Reply by lrljoe** [reply]
So in v3 we are exposing more methods for storing/retrieving table states.

 I'm still looking at the various options, any ideas other than URL/Session/Cookie are welcome
